### PR TITLE
chore(CIS): add protect-kernel-defaults

### LINKED ISSUE #446: kubelet config: --protect-kernel-defaults
We want to deliver the `--protect-kernel-defaults` kubelet option, but are currently unable because we're building clusters without the expected kernel default settings. On a recently built 1.7 cluster:

```
kubelet.go:1323] Failed to start ContainerManager [Invalid kernel flag: vm/overcommit_memory, expected value: 1, actual value: 0, Invalid kernel flag: kernel/panic, expected value: 10, actual value: 0, Invalid kernel flag: kernel/panic_on_oops, expected value: 1, actual value: 0]
```

### PR DESCRIPTION
**Pull request by andyzhangx**
**Reason for Change**:

fix: add protect-kernel-defaults for CIS conformance

about `protect-kernel-defaults`: Default kubelet behaviour for kernel tuning. If set, kubelet errors if any of kernel tunables is different than kubelet defaults.

```
[FAIL] 2.1.6 Ensure that the --protect-kernel-defaults argument is set to true (Scored)
2.1.6 If using a Kubelet config file, edit the file to set protectKernelDefaults: true .
If using command line arguments, edit the kubelet service file
/etc/systemd/system/kubelet.service on each worker node and
set the below parameter in KUBELET_SYSTEM_PODS_ARGS variable.
--protect-kernel-defaults=true
Based on your system, restart the kubelet service. For example:
systemctl daemon-reload
systemctl restart kubelet.service
```

**Issue Fixed**:

Fixes #446

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

/assign @jackfrancis @CecileRobertMichon
cc  @feiskyer

### CODE REVIEW

**Comment by andyzhangx** [comment]
it looks like this flag would make kubelet down:
```
Apr 10 13:07:45 k8s-master-29058287-0 kubelet[23084]: F0410 13:07:45.841909   23084 kubelet.go:1324] Failed to start ContainerManager [Invalid kernel flag: vm/overcommit_memory, expected value: 1, actual value: 0, Invalid kernel flag: kernel/panic, expected value: 10, actual value: 0, Invalid kernel flag: kernel/panic_on_oops, expected value: 1, actual value: 0]
Apr 10 13:07:45 k8s-master-29058287-0 systemd[1]: kubelet.service: Main process exited, code=exited, status=255/n/a
Apr 10 13:07:45 k8s-master-29058287-0 systemd[1]: kubelet.service: Unit entered failed state.
Apr 10 13:07:45 k8s-master-29058287-0 systemd[1]: kubelet.service: Failed with result 'exit-code'.
Apr 10 13:07:46 k8s-master-29058287-0 systemd[1]: kubelet.service: Service hold-off time over, scheduling restart.
```

/hold
Let me hold this PR to do further investigation.

**Comment by andyzhangx** [comment]
close this PR since `protect-kernel-defaults` flag is incompatable with Azure Ubuntu kernel, we won't fix this CIS failure.

**Comment by andyzhangx** [comment]
Interesting thing:
 - In the beginning on VM set up, `/proc/sys/kernel/panic` is 0, so kubelet could not start up with `--protect-kernel-defaults=true` since `/proc/sys/kernel/panic` check failed, and then when I removed the `--protect-kernel-defaults=true` flag manually on that machine, and kubelet could start up, and then `/proc/sys/kernel/panic` changed to 10. This time I could configure kubelet with  `--protect-kernel-defaults=true` flag, all are working well.

```
# cat /proc/sys/kernel/panic
0
$ sudo systemctl daemon-reload
$ sudo systemctl restart kubelet
Warning: kubelet.service changed on disk. Run 'systemctl daemon-reload' to reload units.
$ cat /proc/sys/kernel/panic
10
```

That's the reason why this PR always failed.
cc @jackfrancis @feiskyer

**Comment by jackfrancis** [comment]
@cchildress @khenidak @cpuguy83 Any thoughts on changing the host OS setting `/proc/sys/kernel/panic` to `10`? (from `0`, i.e., disabled)

**Comment by andyzhangx** [comment]
Thanks guys, I have figured out that I should set kernel turnable falgs before set up kubelet, it works now.
The code change only happens on Linux, while there are some Windows test failures.

**Comment by jackfrancis** [comment]
@andyzhangx I've moved bits around here:

1) now validating config via E2E
2) made this a default and not a static (i.e., not user-overridable) configuration
3) added unit tests

I'd still like a review from one of @cchildress @khenidak @cpuguy83 @weinong @yangl900 before merging

File: parts/k8s/cloud-init/artifacts/sysctl-d-60-CIS.conf [hunk 1]
```diff
@@ -18,4 +18,8 @@ net.ipv6.conf.all.accept_ra = 0
 net.ipv6.conf.default.accept_ra = 0
 # 3.3.2 Ensure IPv6 redirects are not accepted
 net.ipv6.conf.all.accept_redirects = 0
-net.ipv6.conf.default.accept_redirects = 0
\ No newline at end of file
+net.ipv6.conf.default.accept_redirects = 0
+# refer to https://github.com/kubernetes/kubernetes/blob/75d45bdfc9eeda15fb550e00da662c12d7d37985/pkg/kubelet/cm/container_manager_linux.go#L359-L397
+vm.overcommit_memory = 1
```
Comment (andyzhangx) [inline parts/k8s/cloud-init/artifacts/sysctl-d-60-CIS.conf]: @jackfrancis I found this code change will fail all e2e test, and actully the kernel flags are not set:
```
Failed to start ContainerManager [Invalid kernel flag: kernel/panic, expected value: 10, actual value: 0, Invalid kernel flag: kernel/panic_on_oops, expected value: 1, actual value: 0, Invalid kernel flag: vm/overcommit_memory, expected value: 1, actual value: 0]
```
There is also no `/etc/sysctl.d/60-CIS.conf` file under `/etc/sysctl.d/`

**Comment by jackfrancis** [comment]
@andyzhangx, we haven't baked this new config file into a VHD yet, so we need to verify this functionality by using the "ubuntu" distro. I've just made a change to do that, and validated that it works:

```
$ sudo cat /etc/default/kubelet 

KUBELET_OPTS=

KUBELET_CONFIG=--address=0.0.0.0 --allow-privileged=true --anonymous-auth=false --authorization-mode=Webhook --azure-container-registry-config=/etc/kubernetes/azure.json --cgroups-per-qos=true --client-ca-file=/etc/kubernetes/certs/ca.crt --cloud-config=/etc/kubernetes/azure.json --cloud-provider=azure --cluster-dns=10.0.0.10 --cluster-domain=cluster.local --enforce-node-allocatable=pods --event-qps=0 --eviction-hard=memory.available<750Mi,nodefs.available<10%,nodefs.inodesFree<5% --feature-gates=PodPriority=true --image-gc-high-threshold=85 --image-gc-low-threshold=80 --image-pull-progress-deadline=30m --keep-terminated-pod-volumes=false --kubeconfig=/var/lib/kubelet/kubeconfig --max-pods=30 --network-plugin=cni --node-status-update-frequency=10s --non-masquerade-cidr=0.0.0.0/0 --pod-infra-container-image=k8s.gcr.io/pause-amd64:3.1 --pod-manifest-path=/etc/kubernetes/manifests --pod-max-pids=-1 --protect-kernel-defaults=true --streaming-connection-idle-timeout=5m 
KUBELET_IMAGE=k8s.gcr.io/hyperkube-amd64:v1.14.1
KUBELET_NODE_LABELS=kubernetes.io/role=master,node-role.kubernetes.io/master=,kubernetes.azure.com/cluster=kubernetes-westus2-11363



KUBELET_REGISTER_NODE=--register-node=true
KUBELET_REGISTER_WITH_TAINTS=--register-with-taints=node-role.kubernetes.io/master=true:NoSchedule
$ kubectl get nodes -o wide
NAME                                 STATUS   ROLES    AGE     VERSION   INTERNAL-IP    EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
k8s-agentpool1-14659950-vmss000000   Ready    agent    3m28s   v1.14.1   10.240.0.4     <none>        Ubuntu 16.04.6 LTS   4.15.0-1042-azure   docker://3.0.4
k8s-agentpool1-14659950-vmss000001   Ready    agent    3m28s   v1.14.1   10.240.0.35    <none>        Ubuntu 16.04.6 LTS   4.15.0-1042-azure   docker://3.0.4
k8s-master-14659950-0                Ready    master   3m28s   v1.14.1   10.255.255.5   <none>        Ubuntu 16.04.6 LTS   4.15.0-1042-azure   docker://3.0.4
```

**Comment by jackfrancis** [comment]
Validated new E2E test coverage on "ubuntu" distro:

```
  should validate Ubuntu host OS network configuration on all nodes
  /Users/jackfrancis/work/src/github.com/Azure/aks-engine/test/e2e/kubernetes/kubernetes_test.go:207

$ k config view -o json

$ k get nodes -o json

$ scp -i /Users/jackfrancis/work/src/github.com/Azure/aks-engine/_output/kubernetes-westus2-11363-ssh -o StrictHostKeyChecking=no scripts/net-config-validate.sh azureuser@kubernetes-westus2-11363.westus2.cloudapp.azure.com:/tmp/net-config-validate.sh
2019/04/25 09:16:45 
Authorized uses only. All activity may be monitored and reported.



$ ssh -A -i /Users/jackfrancis/work/src/github.com/Azure/aks-engine/_output/kubernetes-westus2-11363-ssh -o ConnectTimeout=30 -o StrictHostKeyChecking=no azureuser@kubernetes-westus2-11363.westus2.cloudapp.azure.com -p 22 scp -o StrictHostKeyChecking=no /tmp/net-config-validate.sh k8s-agentpool1-14659950-vmss000000:/tmp/net-config-validate.sh

$ ssh -A -i /Users/jackfrancis/work/src/github.com/Azure/aks-engine/_output/kubernetes-westus2-11363-ssh -p 22 -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR azureuser@kubernetes-westus2-11363.westus2.cloudapp.azure.com ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR k8s-agentpool1-14659950-vmss000000 "/tmp/net-config-validate.sh"

$ ssh -A -i /Users/jackfrancis/work/src/github.com/Azure/aks-engine/_output/kubernetes-westus2-11363-ssh -o ConnectTimeout=30 -o StrictHostKeyChecking=no azureuser@kubernetes-westus2-11363.westus2.cloudapp.azure.com -p 22 scp -o StrictHostKeyChecking=no /tmp/net-config-validate.sh k8s-agentpool1-14659950-vmss000001:/tmp/net-config-validate.sh

$ ssh -A -i /Users/jackfrancis/work/src/github.com/Azure/aks-engine/_output/kubernetes-westus2-11363-ssh -p 22 -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR azureuser@kubernetes-westus2-11363.westus2.cloudapp.azure.com ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR k8s-agentpool1-14659950-vmss000001 "/tmp/net-config-validate.sh"

$ ssh -A -i /Users/jackfrancis/work/src/github.com/Azure/aks-engine/_output/kubernetes-westus2-11363-ssh -o ConnectTimeout=30 -o StrictHostKeyChecking=no azureuser@kubernetes-westus2-11363.westus2.cloudapp.azure.com -p 22 scp -o StrictHostKeyChecking=no /tmp/net-config-validate.sh k8s-master-14659950-0:/tmp/net-config-validate.sh

$ ssh -A -i /Users/jackfrancis/work/src/github.com/Azure/aks-engine/_output/kubernetes-westus2-11363-ssh -p 22 -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR azureuser@kubernetes-westus2-11363.westus2.cloudapp.azure.com ssh -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR k8s-master-14659950-0 "/tmp/net-config-validate.sh"
•
azureuser@k8s-master-14659950-0:~$ grep 'cat /proc/sys/kernel/panic' /tmp/net-config-validate.sh 
cat /proc/sys/kernel/panic | grep $KERNEL_PANIC_VALUE || exit 1
cat /proc/sys/kernel/panic_on_oops | grep $KERNEL_PANIC_ON_OOPS_VALUE || exit 1
azureuser@k8s-master-14659950-0:~$ grep '/proc/sys/vm/overcommit_memory' /tmp/net-config-validate.sh 
cat /proc/sys/vm/overcommit_memory | grep $VM_OVERCOMMIT_MEMORY_VALUE || exit 1
```

File: parts/k8s/cloud-init/artifacts/sysctl-d-60-CIS.conf (hunk 1 above)
Comment (jackfrancis) [reply parts/k8s/cloud-init/artifacts/sysctl-d-60-CIS.conf]: Addressed by putting the whole thing behind "if ubuntu distro" only logic. Once the changes in this file are baked into the VHD then we can turn the functionality on, and run tests, by default

**Pull request merged**